src/localFunction.ts

```diff
@@ -0,0 +1,225 @@
+import * as request from "request";
+
+import { encodeFirestoreValue } from "./firestore/encodeFirestoreValue";
+import * as utils from "./utils";
+import { EmulatedTriggerDefinition } from "./emulator/functionsEmulatorShared";
+import { FunctionsEmulatorShell } from "./emulator/functionsEmulatorShell";
+import { AuthMode } from "./emulator/events/types";
+
+/**
+ * @class
+ * @this LocalFunction
+ * @param {object} trigger
+ * @param {object=} urls
+ * @param {object=} controller
+ */
+export default class LocalFunctionClass {
+  private url?: string;
+
+  constructor(
+    private trigger: EmulatedTriggerDefinition,
+    urls: Record<string, string>,
+    private controller: FunctionsEmulatorShell
+  ) {
+    this.url = urls[trigger.id];
+  }
+
+  private substituteParams(resource: string, params: Record<string, string>) {
+    const wildcardRegex = new RegExp("{[^/{}]*}", "g");
```

Comment:
Nit: consider moving this out of the function so that we don't have to compile it on every call